### PR TITLE
Replace sealed keyword in sealed arrays

### PR DESCRIPTION
**Pull request by Anoukh**
## Purpose
Change the syntax of sealed arrays which used `sealed` keyword to infer size.

### Old Syntax:
```ballerina
sealed int[] sealedArray = [1, 3, 5, 7, 100];
```

### New Syntax:
```ballerina
int[!...] sealedArray = [1, 3, 5, 7, 100];
```

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java [hunk 1]
```diff
@@ -597,13 +597,11 @@ public void visit(BLangArrayType arrayTypeNode) {
...
-                if (arrayTypeNode.isOpenSealed && i == arrayTypeNode.dimensions - 1) {
-                    // Only first dimension is open sealed
-                    resultType = new BArrayType(resultType, arrayTypeSymbol, size, BArrayState.OPEN_SEALED);
-                } else {
-                    resultType = size == -1 ? new BArrayType(resultType, arrayTypeSymbol, size, BArrayState.UNSEALED) :
-                            new BArrayType(resultType, arrayTypeSymbol, size, BArrayState.CLOSED_SEALED);
-                }
+                resultType = (size == -1) ?
+                        new BArrayType(resultType, arrayTypeSymbol, size, BArrayState.UNSEALED) :
+                        (size == -2) ?
```
Comment (hasithaa) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java]: Shall we use constants here.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 2]
```diff
@@ -488,9 +488,15 @@ public void visit(BLangVariable varNode) {
...
-                        "right hand side array literal expected");
+            if (lhsType.tag == TypeTags.ARRAY) {
+                BType type = lhsType;
+                do {
+                    if (((BArrayType) type).state == BArrayState.OPEN_SEALED) {
+                        dlog.error(varNode.pos, DiagnosticCode.INVALID_USAGE_OF_SEALED_TYPE, "array not initialized");
+                        return;
+                    }
+                    type = ((BArrayType) type).eType;
+                } while (type.tag == TypeTags.ARRAY);
```
Comment (SupunS) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java]: Would a recursive method make this simple?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 3]
```diff
@@ -489,9 +489,15 @@ public void visit(BLangSimpleVarRef varRefExpr) {
...
-            return;
+        if (expType.tag == TypeTags.ARRAY) {
+            BType type = expType;
+            do {
+                if (((BArrayType) type).state == BArrayState.OPEN_SEALED) {
+                    dlog.error(varRefExpr.pos, DiagnosticCode.INVALID_USAGE_OF_SEALED_TYPE, "can not infer array size");
+                    return;
+                }
+                type = ((BArrayType) type).eType;
+            } while (type.tag == TypeTags.ARRAY);
```
Comment (SupunS) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]: Similar to above, we can write a recursive method, which i think would be simple and clean.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 4]
```diff
@@ -488,9 +488,15 @@ public void visit(BLangVariable varNode) {
...
         BLangExpression rhsExpr = varNode.expr;
         if (rhsExpr == null) {
-            if (lhsType.tag == TypeTags.ARRAY && ((BArrayType) lhsType).state == BArrayState.OPEN_SEALED) {
-                dlog.error(varNode.pos, DiagnosticCode.INVALID_USAGE_OF_SEALED_TYPE,
-                        "right hand side array literal expected");
+            if (lhsType.tag == TypeTags.ARRAY) {
+                BType type = lhsType;
+                do {
+                    if (((BArrayType) type).state == BArrayState.OPEN_SEALED) {
+                        dlog.error(varNode.pos, DiagnosticCode.INVALID_USAGE_OF_SEALED_TYPE, "array not initialized");
```
Comment (SupunS) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java]: Lets make the error message includes `array not initialized` part as well.

And lets create separate diagnostic codes, for the separate errors, even though some part of the error message is identical.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/Constants.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+ *
+ * since 0.980.0
+ */
+public class Constants {
+
+    private Constants() {
+    }
+
+    public static final int OPEN_SEALED_ARRAY_INDICATOR = -2;
+    public static final int UNSEALED_ARRAY_INDICATOR = -1;
```
Comment (SupunS) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/Constants.java]: Shall we move this to BLangConstants class?

Comment (Anoukh) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/Constants.java]: This class is in ballerina-core and compiler has no dependency to it.

**Pull request merged**